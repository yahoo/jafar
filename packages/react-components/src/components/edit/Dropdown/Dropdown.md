<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar-org/react-form/Form').default;
const Field = require('@jafar-org/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    friends: {
      label: 'Best Friend',
      path: 'bestFriend',
      component: {
        name: 'Dropdown',
        state: {
            items: [{
                value: '1',
                label: 'Ross Geller'
            }, {
                value: '2',
                label: 'Monica Geller'
            }, {
                value: '3',
                label: 'Rachel Green'
            }, {
                value: '4',
                label: 'Chandler Bing'
            }, {
                value: '5',
                label: 'Joey Tribbiani'
            }, {
                value: '6',
                label: 'Phoebe Buffay'
            }]
        }
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      bestFriend: '2', // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    Dropdown: {
      renderer: Dropdown,
    },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='friends' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
    value: 'BASKETBALL',
    state: {
      items: [{
          value: 'BASKETBALL',
          label: 'Basketball'
        }, {
          value: 'FOOTBALL',
          label: 'Football'
        }, {
          value: 'SHOP',
          label: 'Shop'
        }, {
          value: 'FASHION',
          label: 'Fashion'
        }, {
          value: 'COOK',
          label: 'Cook'
        }]
    },
};

 <Dropdown
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
    value: undefined,
    state: {
      items: [{
          value: 'BASKETBALL',
          label: 'Basketball'
        }, {
          value: 'FOOTBALL',
          label: 'Football'
        }, {
          value: 'SHOP',
          label: 'Shop'
        }, {
          value: 'FASHION',
          label: 'Fashion'
        }, {
          value: 'COOK',
          label: 'Cook'
        }]
    },
};
 <Dropdown
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Item's value as any object</h4>

```jsx
const items = [{
          value: {type: 'hobbit', name: 'Frodo'},
          label: 'Frodo'
        }, {
          value: {type: 'hobbit', name: 'Sam'},
          label: 'Sam'
        }, {
          value: {type: 'hobbit', name: 'Pippin'},
          label: 'Pippin'
        }, {
          value: {type: 'elf', name: 'Legolas'},
          label: 'Legolas'
        }, {
          value: {type: 'dwarf', name: 'Gilmli'},
          label: 'Gilmli'
        }]
initialState = { 
    value: {type: 'hobbit', name: 'Pippin'},
    state: {
      items,
    },
};
 <Dropdown
    value={state.value}
    state={state.state}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Required</h4>

```javascript
initialState = { 
    value: 'BASKETBALL',
    state: {
      items: [{
          value: 'BASKETBALL',
          label: 'Basketball'
        }, {
          value: 'FOOTBALL',
          label: 'Football'
        }, {
          value: 'SHOP',
          label: 'Shop'
        }, {
          value: 'FASHION',
          label: 'Fashion'
        }, {
          value: 'COOK',
          label: 'Cook'
        }]
    },
};
 <Dropdown
    value={state.value}
    state={state.state}
    required={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Disabled</h4>

```javascript
initialState = { 
    value: 'FOOTBALL',
    state: {
      items: [{
          value: 'BASKETBALL',
          label: 'Basketball'
        }, {
          value: 'FOOTBALL',
          label: 'Football'
        }, {
          value: 'SHOP',
          label: 'Shop'
        }, {
          value: 'FASHION',
          label: 'Fashion'
        }, {
          value: 'COOK',
          label: 'Cook'
        }]
    },
};
 <Dropdown
    value={state.value}
    state={state.state}
    disabled={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Invalid</h4>

```javascript
initialState = { 
    value: 'FOOTBALL',
    state: {
      items: [{
          value: 'BASKETBALL',
          label: 'Basketball'
        }, {
          value: 'FOOTBALL',
          label: 'Football'
        }, {
          value: 'SHOP',
          label: 'Shop'
        }, {
          value: 'FASHION',
          label: 'Fashion'
        }, {
          value: 'COOK',
          label: 'Cook'
        }]
    },
};
 <Dropdown
    value={state.value}
    state={state.state}
    invalid={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```