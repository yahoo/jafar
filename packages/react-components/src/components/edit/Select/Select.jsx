/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import SelectInternal from 'react-select';
import { isEqual } from 'lodash';
import PropTypes from 'prop-types';

export const styleOverrides = {
  control: {
    fontFamily: 'Roboto, Helvetica, Arial, sans-serif',
    borderTopWidth: '0px',
    borderLeftWidth: '0px',
    borderRightWidth: '0px',
    borderRadius: '0px',
    boxShadow: '0 0 0 0',
    backgroundColor: 'transparent',
    minHeight: '32px',
  },
  singleValue: {
    font: 'inherit',
    padding: '6px 0 7px 0',
    width: '100%',
    margin: 0,
  },
  valueContainer: {
    font: 'inherit',
    padding: '0',
  },
  clearIndicator: {
    padding: '6px 5px 5px 5px',
  },
  dropdownIndicator: {
    padding: '6px 5px 5px 5px',
  },
  menu: {
    fontFamily: 'Roboto, Helvetica, Arial, sans-serif',
  },
};

export const customStyles = {
  control: (provided) => ({
    ...provided,
    ...styleOverrides.control,
  }),
  singleValue: (provided) => ({
    ...provided,
    ...styleOverrides.singleValue,
  }),
  valueContainer: (provided) => ({
    ...provided,
    ...styleOverrides.valueContainer,
  }),
  clearIndicator: (provided) => ({
    ...provided,
    ...styleOverrides.clearIndicator,
  }),
  dropdownIndicator: (provided) => ({
    ...provided,
    ...styleOverrides.dropdownIndicator,
  }),
  menu: (provided) => ({
    ...provided,
    ...styleOverrides.menu,
  }),
};

/**
 * Represent a selection of a single item from fixed items list
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Select/Select.jsx">
 Select</a> from '@jafar-org/react-components/edit/Select'
 */
export default class Select extends React.Component {
  static propTypes = {
    value: PropTypes.any,
    state: PropTypes.shape({
      items: PropTypes.array,
      placeholder: PropTypes.string,
      searchable: PropTypes.bool,
      searchQuery: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    required: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
    onStateChange: PropTypes.func,
  };

  static defaultProps = {
    value: undefined,
    state: {
      items: [],
      placeholder: 'Search',
      searchable: false,
      searchQuery: '',
    },
    disabled: false,
    required: false,
  };

  render() {
    // null - fixes underline controlled VS uncontrolled issue when value turns to undefined
    let selected = (this.props.state.items || []).find(x => isEqual(x.value, this.props.value)) || null;

    return (
      <SelectInternal
        value={selected}
        onChange={this.onValueChange}
        options={this.props.state.items}
        placeholder={this.props.state.placeholder}
        isDisabled={this.props.disabled}
        isClearable={!this.props.required}
        isSearchable={!!this.props.state.searchable}
        inputValue={this.props.state.searchQuery}
        onInputChange={this.onSearchChange}
        styles={customStyles}
      />
    );
  }

  onValueChange = (selected) => {
    let newValue = selected ? selected.value : selected;
    this.props.onValueChange(newValue);
  }

  onSearchChange = (query) => {
    const newState = Object.assign({}, this.props.state);
    newState.searchQuery = query;
    this.props.onStateChange(newState);
  }
}
