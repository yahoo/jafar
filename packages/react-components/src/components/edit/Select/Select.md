<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    bestFriend: {
      label: 'Best Friend',
      path: 'bestFriend',
      component: {
        name: 'Select',
        state: {
        items: [{
                value: '1',
                label: 'Ross Geller'
            }, {
                value: '2',
                label: 'Monica Geller'
            }, {
                value: '3',
                label: 'Rachel Green'
            }, {
                value: '4',
                label: 'Chandler Bing'
            }, {
                value: '5',
                label: 'Joey Tribbiani'
            }, {
                value: '6',
                label: 'Phoebe Buffay'
            }],
        },
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      bestFriend: '2', // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    Select: {
      renderer: Select,
    }, 
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='bestFriend' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
  ```

<h4>Value as object</h4>

```javascript
initialState = { 
  value: { id: '3', first: 'Rachel', last: 'Green' },
  state: {
    items:[{
      value: { id: '1', first: 'Ross', last: 'Geller' },
      label: 'Ross Geller'
    }, {
      value: { id: '2', first: 'Monica', last: 'Geller' },
      label: 'Monica Geller'
    }, {
      value: { id: '3', first: 'Rachel', last: 'Green' },
      label: 'Rachel Green'
    }, {
      value: { id: '4', first: 'Chandler', last: 'Bing' },
      label: 'Chandler Bing'
    }, {
      value: { id: '5', first: 'Joey', last: 'Tribbiani' },
      label: 'Joey Tribbiani'
    }, {
      value: { id: '6', first: 'Phoebe', last: 'Buffay' },
      label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```


<h4>Search</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    searchable: true,
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```

<h4>Required (cant unselect)</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    items:[{
      value: '1',
      label: 'Ross Geller'
    }, {
      value: '2',
      label: 'Monica Geller'
    }, {
      value: '3',
      label: 'Rachel Green'
    }, {
      value: '4',
      label: 'Chandler Bing'
    }, {
      value: '5',
      label: 'Joey Tribbiani'
    }, {
      value: '6',
      label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  required={true}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
  ```
<h4>Disabled</h4>

```javascript
initialState = { 
  value: '3',
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  disabled={true}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```

<h4>Value undefined</h4>

```javascript
initialState = { 
  value: undefined,
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<Select
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```