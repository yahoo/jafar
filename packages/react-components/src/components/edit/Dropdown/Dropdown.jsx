/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import { isEqual } from 'lodash';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';

/**
 * Represent value of any type
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Dropdown/Dropdown.jsx">
 Dropdown</a> from '@jafar-org/react-components/edit/Dropdown'
 */
export default class Dropdown extends React.Component {
  static propTypes = {
    value: PropTypes.any,
    state: PropTypes.shape({
      items: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string.isRequired,
        value: PropTypes.any.isRequired,
      })).isRequired,
      clearText: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    invalid: PropTypes.bool,
    required: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: undefined,
    state: {
      items: [],
      clearText: 'None',
    },
    disabled: false,
    invalid: false,
    required: false,
  };

  getSelectedIndex() {
    const items = this.props.state.items || [];
    return items.findIndex(item => isEqual(this.props.value, item.value));
  }

  getSelected() {
    return this.props.state.items[this.getSelectedIndex()];
  }

  render() {
    if (!(this.props.state.items || []).length) {
      return null;
    }
    const item = this.getSelected();
    const value = item ? JSON.stringify(item.value) : '';

    return (
      <Select
        value={value}
        disabled={this.props.disabled}
        error={this.props.invalid}
        onChange={this.onValueChange}>
        {
          !this.props.required
          && <MenuItem value="">
            <em>{this.props.clearText || Dropdown.defaultProps.state.clearText}</em>
          </MenuItem>
        }
        {this.props.state.items.map((item) => {
          const value = JSON.stringify(item.value);
          return (<MenuItem key={item.label} value={value}>{item.label}</MenuItem>);
        })}
      </Select>
    );
  }

  onValueChange = (e) => {
    let value = e.target.value !== '' ? JSON.parse(e.target.value) : e.target.value;
    value = value === '' ? undefined : value;
    this.props.onValueChange(value);
  }
}
