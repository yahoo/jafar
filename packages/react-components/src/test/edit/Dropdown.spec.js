/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import Select from '@material-ui/core/Select';
import Dropdown from '../../edit/Dropdown/index';


describe('<Dropdown />', () => {
  let component;
  let onValueChangeSpy;
  const value = 'FOOTBALL';
  const state = {
    items: [{
      value: 'BASKETBALL',
      label: 'Basketball',
    }, {
      value: 'FOOTBALL',
      label: 'Football',
    }, {
      value: 'SHOP',
      label: 'Shop',
    }, {
      value: 'FASHION',
      label: 'Fashion',
    }, {
      value: 'COOK',
      label: 'Cook',
    }],
  };
  const disabled = false;
  const invalid = false;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('should trigger onValueChange callback on value change', () => {
    component = mount(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    const newValue = 'FASHION';
    const stringifiedNewValue = JSON.stringify(newValue); // value is stringify in the component
    component.find(Select).at(0).props().onChange({ target: { value: stringifiedNewValue } });
    expect(onValueChangeSpy).toHaveBeenCalledWith(newValue);
  });

  function getComponent(value, state, disabled, invalid, onValueChange) {
    return (
      <Dropdown
        value={value}
        state={state}
        disabled={disabled}
        invalid={invalid}
        onValueChange={onValueChange}
      />
    );
  }
});
