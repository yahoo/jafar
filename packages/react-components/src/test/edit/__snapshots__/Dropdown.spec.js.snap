// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render provided data 1`] = `
<WithStyles(ForwardRef(Select))
  disabled={false}
  error={false}
  onChange={[Function]}
  value="\\"FOOTBALL\\""
>
  <WithStyles(ForwardRef(MenuItem))
    value=""
  >
    <em>
      None
    </em>
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    key="Basketball"
    value="\\"BASKETBALL\\""
  >
    Basketball
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    key="Football"
    value="\\"FOOTBALL\\""
  >
    Football
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    key="Shop"
    value="\\"SHOP\\""
  >
    Shop
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    key="Fashion"
    value="\\"FASHION\\""
  >
    Fashion
  </WithStyles(ForwardRef(MenuItem))>
  <WithStyles(ForwardRef(MenuItem))
    key="Cook"
    value="\\"COOK\\""
  >
    Cook
  </WithStyles(ForwardRef(MenuItem))>
</WithStyles(ForwardRef(Select))>
`;
